### ISSUE TITLE
Cleaning up "testcode" vs "test" vs "samples"

### ISSUE THREAD

**Issue by mrdudz** [cc65/cc65#1224]
As some of you might have noticed, i spent a bit of effort lately on cleaning up the testcode-, test- and samples directory. This ticket should serve as a base for discussing and planning the next step(s).

First of all, the idea was doing this in two steps:

1) make sure all programs present in those directories do actually compile with the current version of the compiler and provide makefiles in each directory to do just that.

this is mostly done now, at least as far as i could do it. all makefiles should be taken as preliminary right now, they might be changed in the next step (or even deleted with the whole directory they are contained in)

that said, there are some programs in the samples/geos directory that i couldnt figure out how to fix them or what to do with them: dialog.c grphstr.c inittab.c menu.c (those don't seem to be working programs, and they dont make a lot of sense to me either) and geosconio.cvt rmvprot.cvt (those look like they worked once, but i dont know GEOS well enough to fix them - someone else might want to look at doing this)

there is also a rather crude makefile in testcode, which i copied from the samples directory and patched up a bit. this one still needs fixing and cleanup (and ideally it will get removed later anyway). it works for building the testcode dir for the time being however :) building the disk images does NOT work correctly right now, mostly because not all files would fit on the disk (at least for c64) - this would have to be changed later to produce two images perhaps.

2) move as much as possible from testcode to either test or samples. ideally when this is done, the testcode directory can be removed. the goal here is that all code can be used by CI, to make sure its always correctly working with the compiler/tools in the tree.

that said, i already moved everything that did somehow fit into test, so we are mostly talking about moving stuff from testcode to samples.

we should discuss and decide what to move where here in this ticket, because thats something i want to do only once :)

2b) if we don't manage to clear out testcode completely, we should clearly define what kind of stuff belongs into which directory. right now it seems quite arbitrary to me

Questions from my side:
- how do we want to deal with sub-directories in samples (and testcode, should we not manage to remove it). in my own projects i would typically have a makefile in every directory level, which then calls (all) the makefiles in the contained directories. do we want this? (right now we do this in test, but not in samples)
- how to deal with the "boilerplate" stuff in makefiles that deals with the cmd.exe vs sh differences? do we want to minimize that stuff per makefile, or keep a common piece of "code" that is the same everywhere?
- what programs in testcode are completely useless and should be removed? right now to me only the few GEOS samples mentioned above (dialog.c grphstr.c inittab.c menu.c) come to mind - but perhaps there are more

(as a sidenote: my main development box is running linux, if you notice something not working in cmd.exe, let me know. one thing oliver pointed out already is the $(RM) macro, that should be replaced by $(DEL). i will do this in step 2, if any of those makefiles survive :))

**Comment by mrdudz** [comment]
I'll start with suggesting:

testcode/assembler -> move to test/asm
testcode/disasm  -> move to samples/dasm
testcode/grc -> move to samples/geos

that leaves testcode/lib, which will take a bit more time and effort :)

**Comment by oliverschmidt** [comment]
I see a use of three directories (which fortunately maps very nicely to what we already have).

1. samples: Code interesting to cc65 users for learning / understanding cc65 aspects. Part of cc65 deliverables. Built as part of CI.

2. test: Code to verify the correct operation of the cc65 programs (aka binaries). Built as part of CI. Run as part of CI using sim65.

3. libtest (analog to 'src' vs 'libscr'): Code to verify the correct operation of the cc65 libraries. Built as part of CI. **_Not_** run as part of CI for two reasons: a) We're not going to install emulators as part of the CI. b) Many tests aren't to be automated with reasonable amount of effort.

Looking at what we have today it's obvious that just renaming 'testcode' to 'libtest' brings us already pretty close to the picture above.

So samples and libtest are different from the semantic POV and the delivery POV. But I'd like to see them similar from the build POV. I'd like to have...

* Makefiles for (ideally) all code sitting in both directories

* the Makefiles all follow a common style

* the Makefiles all have an default target named 'all' which builds all binaries.

* the Makefiles all have a 'disk' (or 'disks' ?) target which builds disk images. This target requires the same set of disk image managing tools installed for all Makefiles and made known to the Makefile in the same way.

The idea is that the default targets can be built as part of CI while the 'disk(s)' target simplifies interactive execution of the code in the local dev setting.

The primary open question I see is this: On the one hand side I'd like to see rather easy-to-understand Makefiles in the samples directory (and as I said above I'd like to have similar Makefile in samples and libtest that's true for libtest too). On the other hand I'd like to see both samples and libtest not only built for the C64. But doing that in a maintainable fashion tends to result in Makefiles like https://github.com/cc65/cc65/blob/master/samples/Makefile which I presume most people don't consider easy-to-understand.

**Comment by oliverschmidt** [comment]
> testcode/assembler -> move to test/asm
> testcode/grc -> move to samples/geos

ACK

> testcode/disasm -> move to samples/dasm

I have no idea what this code wants to archive in the first place.

**Comment by mrdudz** [comment]
i see the intend of "libtest" - but this is what created the mess we are trying to solve in the first place :) because:

- some "lib" tests can - and because of that should - be run automatically. and they are already in "test". moving them back to "libtest" feels wrong, and like a step backwards.
- stuff in "samples" fulfils the same requirements (will be built, but not run by CI), and many (if not most) of these programs currently in "testcode/lib" are OK to demonstrate certain aspects of the libs too (more than some things currently in "samples" IMHO).

because of those reasons, i find it very hard to come up with a generic definition of what goes where - and that is always a sign of something being wrong. there should be a clear and simple rule we can use to tell what goes where, with just "test" and "samples" that is easy to do. when adding "libtest" it becomes complicated and messy, because of the reasons above.

i see no problem with those programs being moved to "samples" either (perhaps that dir will need a bit more structuring when it gets fuller).

**Comment by mrdudz** [comment]
> > testcode/disasm -> move to samples/dasm
> 
> I have no idea what this code wants to archive in the first place.

it's a simple demonstration of how to use the disassembler - no more no less :)

that question indicates another thing we should probably do though - put a readme into each directory with a brief explaination of what each programs is.

**Comment by oliverschmidt** [comment]
> some "lib" tests can - and because of that should - be run automatically. and they are already in "test".

Please name/list them.

**Comment by mrdudz** [comment]
> > some "lib" tests can - and because of that should - be run automatically. and they are already in "test".
> 
> Please name/list them.

just a quick look at test/val (i might have missed some):

lib_common_atoi.c
lib_common_ctype.c
lib_common_memmove.c
lib_common_mulxx.c
lib_common_strcat.c
lib_common_strchr.c
lib_common_strcspn.c
lib_common_strncat.c
lib_common_strol.c
lib_common_strrchr.c
lib_common_strspn.c
lib_common_strtoul.c
atoi-test.c
rand.c
snprintf-test.c
strncmp-test.c
strnicmp-test.c
strtol-test.c
strtoul-test.c
strpbrk-test.c
time.c
signal-test.c

(interestingly, there seem to be various duplicated tests among those too)

**Comment by mrdudz** [comment]
oh and before i forget, one rule i'd like to see for "samples" (this is not strictly followed yet for "testcode" iirc) would be: "programs in the root of the "samples" directory must be portable and work on different targets. target specific stuff should live in a subdirectory with the name of the target"

**Comment by oliverschmidt** [comment]
I don't see how any of the following could/should be made either CI-testable nor even remotely classifies as a sample:

https://github.com/cc65/cc65/blob/master/testcode/lib/atari/charmapping.c
https://github.com/cc65/cc65/blob/master/testcode/lib/atari/defdev.c
https://github.com/cc65/cc65/blob/master/testcode/lib/atari/displaylist.c
https://github.com/cc65/cc65/blob/master/testcode/lib/atari/mem.c
https://github.com/cc65/cc65/blob/master/testcode/lib/atari/ostype.c
https://github.com/cc65/cc65/blob/master/testcode/lib/atari/scrcode.s
https://github.com/cc65/cc65/blob/master/testcode/lib/atari/sys.c
https://github.com/cc65/cc65/blob/master/testcode/lib/cbm/petscii.c
https://github.com/cc65/cc65/blob/master/testcode/lib/gamate/ctest.c
https://github.com/cc65/cc65/blob/master/testcode/lib/pce/conio.c
https://github.com/cc65/cc65/blob/master/testcode/lib/cpeek-test.c
https://github.com/cc65/cc65/blob/master/testcode/lib/cprintf.c
https://github.com/cc65/cc65/blob/master/testcode/lib/cursor.c
https://github.com/cc65/cc65/blob/master/testcode/lib/em-test.c
https://github.com/cc65/cc65/blob/master/testcode/lib/exec-test1.c
https://github.com/cc65/cc65/blob/master/testcode/lib/exec-test2.c
https://github.com/cc65/cc65/blob/master/testcode/lib/ft.c
https://github.com/cc65/cc65/blob/master/testcode/lib/mul-test.c
https://github.com/cc65/cc65/blob/master/testcode/lib/rename-test.c

**Comment by oliverschmidt** [comment]
> oh and before i forget, one rule i'd like to see for "samples" (this is not strictly followed yet for "testcode" iirc) would be: "programs in the root of the "samples" directory must be portable and work on different targets. target specific stuff should live in a subdirectory with the name of the target"

ACK

**Comment by oliverschmidt** [comment]
>>> some "lib" tests can - and because of that should - be run automatically. and they are already in "test".

>> Please name/list them.

> just a quick look at test/val (i might have missed some):
lib_common_atoi.c [...]

Maybe there's - at least partly - a misunderstanding: I tend to associate "lib" with all the different target-specific library code, not with the portable code in /common and /runtime. I fully agree that the latter could and should be tested using sim65 as part of CI.

Maybe it would help if the term "libtest" in my statement above is replaced with "targettest".

**Comment by mrdudz** [comment]
Sorry for no feedback... i was busy with some other stuff. But i am still at it :) Just trying to catch up with merged PRs and add some more tests...

**Comment by mrdudz** [comment]
so should i start with:

> testcode/assembler -> move to test/asm
> testcode/disasm -> move to samples/dasm
> testcode/grc -> move to samples/geos

?

Also, if "libtest" is supposed to be used for testing only library things that are target specific (obvious example: TGI tests), then wouldnt a less ambigous name a better idea? ("targettest"?)

**Comment by mrdudz** [comment]
another suggestion:
samples/README -> rename to samples/readme.txt

**Comment by mrdudz** [comment]
oh and, another step once all of the above is done:
move all of testcode/lib/* to testcode/ (one level up), then
rename testcode to targettest

after that we can start cleaning up targettest and eventually move some things to samples
@oliverschmidt should i proceed? :)

**Comment by oliverschmidt** [comment]
After carefully re-reading the thread so far it seems to me that we have come to the point of a similar-enough understanding :-)

Please proceed as proposed/discussed - and thanks for your effort!

**Comment by mrdudz** [comment]
OK - done, now we are left with:

a) hook up targettest/Makefile with CI
b) eventually move some stuff from targettest to samples or test

i will proceed with b) over time... it takes a bit of work in some cases to rework the libc related tests to work for CI.
I have no idea how a) would even work, so thats something for you :) 

How do we handle subdirectories within samples or targettest btw, should the toplevel makefile recurse into them and build all of it? (that's what i would usually do). Else, will you add all the subdirs into CI manually (and does that mean when a new one is added, the CI makefile needs to be updated?)

**Comment by oliverschmidt** [comment]
> b) eventually move some stuff from targettest to samples or test [...] i will proceed with b) over time... it takes a bit of work in some cases to rework the libc related tests to work for CI.

I see.

> a) hook up targettest/Makefile with CI [...] I have no idea how a) would even work, so thats something for you :)

I explicitly don't want to see that happen.

> How do we handle subdirectories within samples or targettest btw, should the toplevel makefile recurse into them and build all of it? (that's what i would usually do).

Given that there's https://github.com/cc65/cc65/blob/master/test/Makefile I generally don't see a reason to proceed differently for samples/. However I don't see you haveing fun with extending https://github.com/cc65/cc65/blob/master/samples/Makefile without breaking its approach/style.

**Comment by oliverschmidt** [comment]
> PS: please have a look at test/asm/Makefile and let me know what i did is how we want to do it :)

Thanks for the hint. Actually it is neither the way we want to do it nor actually working as expected. See https://github.com/cc65/cc65/commit/4acdc9ced9ded4c37e46d2bc531bea31894c1ccc

**Comment by mrdudz** [comment]
> I explicitly don't want to see that happen.

now you lost me. i thought we agreed that those programs will be compiled, but not run, in the CI phase. Its the whole point of this reorganizing in the first place. *shrug* (see point 3. in your previous post)

> However I don't see you haveing fun with extending https://github.com/cc65/cc65/blob/master/samples/Makefile without breaking its approach/style.

That's why i am asking. If this would be my project, i would totally rewrite this makefile into something much more explicit, as i find all this implicit stuff way too confusing and complex :) So i am not sure what you a trying to say there...

> Thanks for the hint. Actually it is neither the way we want to do it nor actually working as expected.

neither is your fix, IMHO - now it implies that every .s file is only compiled, and never results in a program that is to be run and the result checked (which is what i'd expect the generic/default rule to do in this directory)

**Comment by oliverschmidt** [comment]
> i thought we agreed that those programs will be compiled, but not run, in the CI phase.

My fault, you're of course right!

> neither is your fix, IMHO

At least the Makefile now understands that things are done while you version tried in vain to get things "done".

> now it implies that every .s file is only compiled, and never results in a program that is to be run and the result checked (which is what i'd expect the generic/default rule to do in this directory)

Not at all. What makes you think so? Line https://github.com/cc65/cc65/blob/master/test/asm/Makefile#L41 explicitly asks for a .o file. Therefore only an .o file is produced. If someone wants a program he'll have to ask for (e.g.) a .bin file (and a rule has to be added to produce one).

**Comment by oliverschmidt** [comment]
> If this would be my project, i would totally rewrite this makefile into something much more explicit, as i find all this implicit stuff way too confusing and complex :)

I know very well as you pointed this out many times by now.

> So i am not sure what you a trying to say there...

...that it looks like I'll have to make the necessary adjustments to that Makefile - here you go: https://github.com/cc65/cc65/commit/fd208fdf0b6f968ba7491dcec1f53be71af08abd

I additionally made the minimal adjustments to /tutorial/Makefile to make it work. I'd appreciate if you would make similar adjustments to the other Makefiles you created in the recent past.

https://github.com/cc65/cc65/blob/master/samples/Makefile#L156 is the point to add additional subdirectories.

**Comment by mrdudz** [comment]
> Not at all. What makes you think so? Line https://github.com/cc65/cc65/blob/master/test/asm/Makefile#L41 explicitly asks for a .o file. Therefore only an .o file is produced. If someone wants a program he'll have to ask for (e.g.) a .bin file (and a rule has to be added to produce one).

right. as you can see, the implicit stuff is confusing for me :)

**Comment by mrdudz** [comment]
Ok, fixed the makefiles and compiling of the whole tree

one question... the disasm example uses and requires cpp - that is no problem for the CI (i guess), but will lead to errors when someone tries to compile the samples directory without having it. so for now it's not added to the list of subdirs. the question is... can we handle it gracefully (so it will be used for CI, but cause no error when no cpp exists)? or do we leave it like this?

**Comment by oliverschmidt** [comment]
> Ok, fixed the makefiles and compiling of the whole tree

Thanks. However, there are some aspects I'd like to be cleaned up:

* https://github.com/cc65/cc65/blob/master/targettest/gamate/Makefile#L42
Maybe it would be nicer to name that `testl` for consistency with `testc` and `testn`.

* https://github.com/cc65/cc65/blob/master/targettest/gamate/Makefile#L44
Maybe it would be nicer to name that `testa` for consistency with `testc` and `testn`.

* https://github.com/cc65/cc65/blob/master/targettest/gamate/Makefile#L43
https://github.com/cc65/cc65/blob/master/targettest/gamate/Makefile#L45
https://github.com/cc65/cc65/blob/master/targettest/gamate/Makefile#L47
https://github.com/cc65/cc65/blob/master/targettest/gamate/Makefile#L49
I don't see our Makefile making presumptions on directories like `~/Desktop/mame/winmess/`.
In general I rather don't see emulators started from Makefiles. However, if you insist to do so, then please
  1. Follow the approach used in https://github.com/cc65/cc65/blob/master/samples/Makefile#L86-L94 for disk management tools (macro set with `?=` allowing the user to override e.g. with an absolute path, comment where to get the tool from)
  2. Don't depend on a certain platform (e.g. set a macro `MESS` based on the value of `CMD_EXE` either to `mess` or `wine mess.exe`).

* https://github.com/cc65/cc65/blob/master/targettest/pce/Makefile#L51
We don't add dependencies on binaries nor libraries to samples or tests. Rather we presume binaries and libraries to "be there".

* https://github.com/cc65/cc65/blob/master/targettest/pce/Makefile#L52
Make use of the `CL` set above explicitly for that purpose.

* https://github.com/cc65/cc65/blob/master/targettest/pce/Makefile#L58
See my comments above regarding MESS.

**Comment by mrdudz** [comment]
ah yes, that "test" stuff shouldnt be there, perhaps... i'd actually prefer to remove this stuff from the makefiles, and instead add a line or two to the documentation, like i apparently already did in https://cc65.github.io/doc/gamate.html

**Comment by oliverschmidt** [comment]
>> I rather don't see emulators started from Makefiles.

> i'd actually prefer to remove this stuff from the makefiles [...]

So we agree here :-)

**Comment by mrdudz** [comment]
So, should i put similar hints into the respective doc pages?

**Comment by oliverschmidt** [comment]
Are you aware of https://cc65.github.io/doc/intro.html#toc6 ?

I propose that you put the content in question in new sections there (and link to them). Maybe it would be nice to have such links for the other targets too. E.g. a link from https://cc65.github.io/doc/apple2.html#s9 and https://cc65.github.io/doc/apple2enh.html#s9 to https://cc65.github.io/doc/intro.html#ss6.1.

Note: Please replace http URLs with https URLs for sites which forward http to https anyhow (as e.g. `http://www.mess.org/` and `http://en.wikipedia.org/wiki/Gamate` do). I've already done that in the past with many URLs (e.g. https://github.com/cc65/cc65/commit/58bfe28244d1edc96d73c43dd9716ee05fafad5f).

**Comment by mrdudz** [comment]
> Are you aware of https://cc65.github.io/doc/intro.html#toc6 ?

I find this page a rather unfortunate place to put this information... as i understand it, target specific stuff does not fit there, it's unrelated to "compiler intro". actually, if i added all that stuff, the pages primary content would be "running executables on all the supported targets".

If it really shouldn't be on the target specific page(s) (that would still be my preference, thats where i'd look in the first place), wouldnt moving this stuff to a new, seperate, page make more sense?

**Comment by oliverschmidt** [comment]
> If it really shouldn't be on the target specific page(s) (that would still be my preference, thats where i'd look in the first place), wouldnt moving this stuff to a new, seperate, page make more sense?

I'm pretty much with you, but there's one section for all CBMs and one section for both Apple2's. How to move that to the existing target pages?

**Comment by polluks** [comment]
I know what you mean, it's more emulator related but some kind of first aid.

**Comment by mrdudz** [comment]
>  but there's one section for all CBMs and one section for both Apple2's. How to move that to the existing target pages?

good point. so a seperate page? :)

**Comment by mrdudz** [comment]
77bfd163cd2460b46d4e31b17eb5e5e7ec6ab8aa cleans up the makefiles. i'm going through the target pages now and move the emulator stuff into the "intro" page for the time being, so the only question left is whether we want it on a seperate page.

**Comment by mrdudz** [comment]
hrm, now i ran into an annoying problem. My emulator of choice for PCE and NES is "mednafen". But there apparently cant be two "sections" with that name on the doc page. How to deal with this? :)

**Comment by greg-king5** [comment]
Hyphen-names, maybe.  `PCE-Mednafen` and `NES-Mednafen`.

**Comment by oliverschmidt** [comment]
> wouldnt moving this stuff to a new, seperate, page make more sense?

> so a seperate page? :)

I'm totally with you. A separate page regarding the workflow of getting from the ld65 output file to the the actually running program for several target(-groups) / emulators would likely be the best approach.

**Comment by oliverschmidt** [comment]
If a single emulator is used / can be used for several targets and the workflow is (almost) identical for the targets I personally don't consider it necessary (or in fact even desirable) to have multiple (almost) identical documentation sections.

**Comment by mrdudz** [comment]
> If a single emulator is used / can be used for several targets and the workflow is (almost) identical for the targets I personally don't consider it necessary (or in fact even desirable) to have multiple (almost) identical documentation sections.

IMHO it'd be very confusing if we didnt just repeat those 3 lines

**Comment by oliverschmidt** [comment]
Of course, if it's just three lines...

**Comment by mrdudz** [comment]
Not sure why this was closed :) i still have some things pending (mostly doc changes) related to this..... (yeah i know, it has been a while)

I also noticed some other open ends too that probably should get fixed, but i am not sure how exactly (you know the travis stuff better than i do):

- apparently the "targettest" directory is not compiled at all when testing automatically (a "make -C targettest all" missing in the travis.yml?) ... which kinda defeats the point of this whole cleanup :)
- compiling the "targettest" directory leaves a lot of clutter behind (intermediate files and executables), which is annoying to say at least, it would be nice to gitignore that stuff but i am not sure what the best way to achive this would be. Putting all files into "wrk" doesnt seem to be ideal, because someone might want to run the executables or use them as examples.
- it would be useful if the toplevel "test" target also triggered compiling the "targettest" stuff, so running those wont be forgotten. or atleast if a seperate "targettest" target existed that triggers "make -C targettest all".

- the "samples" directory has a similar problem with the clutter, but obviously here we dont want the files elsewhere. if we could agree on using the common .prg extension for all CBM files, that would at least allow to gitignore most of the clutter with a few wildcards (same for the "targettest" dir, i guess) (actually some subdirs use them, some do not... which is odd. my vote is on using it everywhere)

also:
- (mostly a note to myself) there should probably be a short readme in "targettest" that explains what exactly should be there (derived from the discussion here). similar thing in "samples", i guess

**Comment by oliverschmidt** [comment]
I don't understand this section of `targettest/Makefile`:

```
testcode: $(EXELIST_$(SYS))

all: testcode
	make -C accelerator
	make -C apple2
	make -C atari
	make -C atari5200
	make -C cbm SYS=$(SYS)
	make -C gamate
	make -C pce
```

Isn't it desirable to have one target that builds both the tests in the targettest directory as well as in the subdirectories? And I suggest that this target is the first one. As far as I can see `samples/Makefile` already does this very thing...

You suggest to execute `make -C targettest all`. This fails for me with

```
dd if=conio.bin bs=8K skip=1 > conio.pce
'dd' is not recognized as an internal or external command,
operable program or batch file.
```

`dd` isn't part of the list of utilities in https://www.gnu.org/software/make/manual/html_node/Utilities-in-Makefiles.html#Utilities-in-Makefiles. I suggest to move that build step to another target which isn't supposed to be "normally" called.

Regarding the files created in targettest I think that ignoring them doesn't make them go away. We have `wrk`, `libwrk` and `testwrk`. It seems pretty natural to me to have a `targettestwrk`. You wrote: "Putting all files into "wrk" doesnt seem to be ideal, because someone might want to run the executables or use them as examples.". If a program is a sample, it should be moved into samples. If it isn't a sample then I don't see a reason to make a different between `test` and `targettest`.

You wrote "it would be useful if the toplevel "test" target also triggered compiling the "targettest" stuff, so running those wont be forgotten.". I don't agree. We currently intentionally don't have cross references between the Makefiles in the different directories. Apart from that, `targettest` and `test` seem to be pretty different, given that the former only builds, but doesn't run any programs.

You wrote "actually some subdirs use them, some do not... which is odd."

- test/.gitignore is from my perspective a leftover and should be removed.
- src/.gitignore is solely for MS VC, which is a very specific case not comparable to anything else in this repo.
- samples/disasm/.gitignore is something I can't comment on. I don't know (anymore?) what this directory is about and why this is considered a sample (for what?).

So all in all I don't see local .gitignore files being a general pattern in this repo. Nevertheless it could be appropriate for samples as (as you write) this is the (from my POV only) case where the output files belong here.

You wrote "using the common .prg extension for all CBM files, that would at least allow to gitignore most of the clutter". I hope you still agree that cc65 isn't only about CBM machines, just because of the `SYS ?= c64` in the Makefile ;-)

**Comment by mrdudz** [comment]
> Isn't it desirable to have one target that builds both the tests in the targettest directory as well as in the subdirectories? And I suggest that this target is the first one. As far as I can see samples/Makefile already does this very thing...

that was the intention - i am however used to "make all" doing just that (and i always type "make all" rather than "make", which is why i didnt notice), and expected the Makefile in samples doing it, which is why put the added stuff there (that Makefile is pretty much a direct copy from the one in samples). But sure, i can fix this :)

> dd isn't part of the list of utilities in https://www.gnu.org/software/make/manual/html_node/Utilities-in-Makefiles.html#Utilities-in-Makefiles. I suggest to move that build step to another target which isn't supposed to be "normally" called.

yes, perhaps. "gamate" probably has the same problem. i tend to forget that there are actually people that dont have those essential tools available in their setups, i cant work without them :) I think we talked about making a little tool for this some day. or add some features to ld65 to support it directly.... another thing TODO :)

> If a program is a sample, it should be moved into samples. If it isn't a sample then I don't see a reason to make a different between test and targettest.

The issue with the programs in targettest is that they mostly dont qualify as a useful sample program (they are either too trivial or do funny stuff that isnt really useful) but still someone might want to check their output to see if they are working - which is why putting the executables somewhere else is a bit annoying. Remember we put them into a seperate directory because we dont want to involve emulators for the automatic testing. However, the direct consequence of that is that (occasionally) someone will run the programs using the emulator manually, and then having the executables somewhere else isnt very useful at all.

However, the problem with ignoring created executables applies to both samples and targettest in very much the same way anyway. Ideally all the local gitignore files can be removed and a couple extra lines in the global one deals with all of it. Putting one into samples and then doing the same thing in targettest seems pointless :)

> You wrote "it would be useful if the toplevel "test" target also triggered compiling the "targettest" stuff, so running those wont be forgotten.". I don't agree. We currently intentionally don't have cross references between the Makefiles in the different directories. Apart from that, targettest and test seem to be pretty different, given that the former only builds, but doesn't run any programs.

OK. I dont understand whats the intend of this i have to admit (eg i frequently have to do "make clean all" because compiler changes do not result in the library being rebuild, and i dont want a mix of new compiler and libs built with old compiler - ever), i expect the toplevel to handle this for me, if i want to build just a subdir i'd explicitly use the makefile from the subdir (and that hardly ever happens)

Anyway, then at least it would be useful to have a "targettest" target in it, so i can do "make clean all test targettest" instead.

> You wrote "actually some subdirs use them, some do not... which is odd."

that referred to CBM programs using the .prg extension, not .gitignore files :)

> samples/disasm/.gitignore is something I can't comment on. I don't know (anymore?) what this directory is about and why this is considered a sample (for what?).

see https://github.com/cc65/cc65/issues/1224#issuecomment-683313058 :)

> I hope you still agree that cc65 isn't only about CBM machines, just because of the SYS ?= c64 in the Makefile ;-)

sure. all i am saying is that using the .prg extension for CBM programs is common. and doing it will allow to gitignore *.prg globally and remove a lot of clutter with a single line in gitignore :) if we dont want to use .prg, then doing the same will require adding all and every filename to be ignored... which is a bit silly, imho.

**Comment by oliverschmidt** [comment]
> that was the intention - i am however used to "make all" doing just that (and i always type "make all" rather than "make", which is why i didnt notice), and expected the Makefile in samples doing it, which is why put the added stuff there (that Makefile is pretty much a direct copy from the one in samples). But sure, i can fix this :)

Maybe there's a misunderstanding: Have you noticed, that the samples/Makfile 'all' target is intentionally(!) empty? See
https://github.com/cc65/cc65/commit/4dcfc036c8d29a9c57f3181931448027aeea0ce4
https://github.com/cc65/cc65/commit/c9e9679a06afd86fe8caeb32177c11736b98a974

> I think we talked about making a little tool for this some day. or add some features to ld65 to support it directly.... another thing TODO :)

* I personally think that it doesn't belong more to ld65 than let's say creation of a .d64 file - but that's another story.
* I agree with you that this is a TODO. Until it isn't solved the creation of the "final" binary for those targets can't be part of a default and/or 'all' target.

> However, the problem with ignoring created executables applies to both samples and targettest in very much the same way anyway.

I pretty much disagree here. The samples folder tries to accomodate (unexperienced) user expecations. Therefore files are generated right there. The targettest folder in contrast is to be used by a pro - and pros never ever have their toolchain mess up their source directory.

> Anyway, then at least it would be useful to have a "targettest" target in it [...]

I don't oppose in any way to add suport for targettest to the main Makefile - as soon as the Makefile works as desired ;-)

I'm certainly not interested in adding any file type mask to the main .gitignore file. And there are targets with not-so-clearly-defined file types. And the samples/Makefile is already (more than) complex enough. Making it add target-specific file type extensions doesn't help ;-)

Okay, let my try to get more productive - I can imagine this:

The build results aren't put it "some" sibling directory, but in a subdirectory. And given that we have several targets with (currently) not disjunct file names, we could have one subdirectory for every target. Or a `build` or `bin` directory containing those subdirectories.

This approach could from my perspective be applied to both samples and targettest.

**Comment by mrdudz** [comment]
> Maybe there's a misunderstanding: Have you noticed, that the samples/Makfile 'all' target is intentionally(!) empty?

Since this is at least.. lets say uncommon, there should at least be a comment saying WHY this is the case... i am not sure i understand the intention for that matter.

> The samples folder tries to accomodate (unexperienced) user expecations. Therefore files are generated right there. The targettest folder in contrast is to be used by a pro - and pros never ever have their toolchain mess up their source directory.

I'd put consistency over all regarding this. One should behave like the other - or it becomes really surprising for one or the other. (and i have seen enough "pros" to not fall for this =))

Will try to fix up those makefiles a bit first... i am not sure if i can change them to build in a seperate directory though, i'd have to rewrite them in a completely different way to do this.

**Comment by oliverschmidt** [comment]
> [...] i am not sure i understand the intention for that matter.

The idea is that a plain top level `make` doesn't build the samples as they aren't an obligatory part of a working cc65 installation. See https://github.com/cc65/cc65/commit/c9e9679a06afd86fe8caeb32177c11736b98a974 for details.

> I'd put consistency over all regarding this. [...]

Doesn't my recent proposal offer (cited below) this very consistency?

> I can imagine this:

> The build results aren't put it "some" sibling directory, but in a subdirectory. And given that we have several targets with (currently) not disjunct file names, we could have one subdirectory for every target. Or a build or bin directory containing those subdirectories.

> This approach could from my perspective be applied to both samples and targettest.

Or are you aming for both the samples and targettest Makefile to be consistent with the rest of the Makefiles?

> Will try to fix up those makefiles a bit first... i am not sure if i can change them to build in a seperate directory though, i'd have to rewrite them in a completely different way to do this.

I fear that if you rewrite them completely they will be completely inconsistent with the other Makefiles ;-)

At least the other Makefiles already build in a separate directory using the currenty Makefile "style". Or are you thinking about changing aspects of the two Makefiles which are not found in the other Mafilefiles too?

**Comment by mrdudz** [comment]
> The idea is that a plain top level make doesn't build the samples as they aren't an obligatory part of a working cc65 installation. See c9e9679 for details.

I looked and no offence, none of it makes sense to me :) You can not dive into a subdirectory no problem and still have a working "all" in the subdir. Is this an artefact of the toplevel "all" calling the "all" target in all subdirs?

> Doesn't my recent proposal offer (cited below) this very consistency?

It does. I just wanted to express that whatever the solution is - it should be the same in both.

> Or are you aming for both the samples and targettest Makefile to be consistent with the rest of the Makefiles?

Well, that too - thats why i copied the one in targettest from samples :)

> I fear that if you rewrite them completely they will be completely inconsistent with the other Makefiles ;-)

definetly. Mine are usually the exact opposite, very explicit, no implied rules etc

> At least the other Makefiles already build in a separate directory using the currenty Makefile "style". Or are you thinking about changing aspects of the two Makefiles which are not found in the other Mafilefiles too?

no i am not implying anything like that. i will only fix as much as i can do (and mostly the *others*, ie in the subdirectories of targettest and samples). someone else will have to do the rest :)

**Comment by mrdudz** [comment]
oh btw, using "echo" is ok (and will work without extra magic on windows?) i'd like to output a message that says what the additional targets are for building images (like the one containing the dd stuff) in the standard targets of the individual makefiles.

**Comment by audetto** [comment]
I think the `geos` folder should be removed from the makefile from `samples`.

```
make SYS=apple2
../../bin/cl65 -t apple2 -O -o bitmap-demo.cvt -m bitmap-demo.map bitmap-demores.grc bitmap-demo.c
grc65: Unsupported target system 'apple2'
make[1]: *** [Makefile:55: bitmap-demo.cvt] Error 1
```

The valid targets for the `samples` folder and `geos` do not intersect.

It only works for a plain `make` because each makefile sets a default `SYS` (`c64` ad `geos-cbm`) which works.
But as soon as one sets `SYS` even `SYS=c64` the `goes` folder cannot compile.

And if I do `make SYS=geos-cbm` then it is the `samples` folder which does not compile.

**Comment by mrdudz** [comment]
@audetto try again, it should work now

**Comment by oliverschmidt** [comment]
@mrdudz: I'm not sure that your approach matches @audetto's expectation.

I suggest that the samples/geos Makefile contains logic to change the value of SYS from `c64` to `geos_cbm` and from `apple2enh` to `geos_apple` iff SYS was given on the command line (using the $(origin ...) function (https://www.gnu.org/software/make/manual/html_node/Origin-Function.html#Origin-Function).

**Comment by mrdudz** [comment]
> I suggest that the samples/geos Makefile contains logic to change the value of SYS from c64 to geos_cbm and from apple2enh to geos_apple iff SYS was given on the command line (using the $(origin ...) function

i can try :)

right now i am struggling to find a good solution for the "clean" target - right now it requires setting SYS= correctly, which is quite unexpected and leads to surprising behaviour

**Comment by audetto** [comment]
> @mrdudz: I'm not sure that your approach matches @audetto's expectation.
> 

You might be reading into my expectations more than there actually is :-)

> I suggest that the samples/geos Makefile contains logic to change the value of SYS from `c64` to `geos_cbm` and from `apple2enh` to `geos_apple` iff SYS was given on the command line (using the $(origin ...) function (https://www.gnu.org/software/make/manual/html_node/Origin-Function.html#Origin-Function).

I think the current solution solves my initial issue. I can now create a apple2 disk easily.

As you might have understood, I a bit new to cc65 and at the moment I don't even know how to use a geos-apple binary.
I understand the pain in switching the target, and I am not asking it.

**Comment by oliverschmidt** [comment]
@mrdudz:

Further above you wrote:

> i will only fix as much as i can do (and mostly the others, ie in the subdirectories of targettest and samples). someone else will have to do the rest :)

I'm still not clear what you expressed back then.

Now you wrote: 

> right now i am struggling to find a good solution for the "clean" target - right now it requires setting SYS= correctly, which is quite unexpected and leads to surprising behaviour

I wrote before that I have the strong opinion, that all generated files should go into seperate (sub-)directories. Then a "clean" just means to empty such directories (or recursively remove them).

Do we agree on that approach but you want me to implement it - or do we disagree on that approach and you are still looking for solutions to workaround the from-my-perspective-broken-in-the-first-place current approach?

**Comment by mrdudz** [comment]
> Do we agree on that approach

i dont like it, but if you prefer it - why not :)

Thats not the point though - i simply like things working at all times, and i dont mind implementing a fix that will be later thrown out and replaced by something else. Now that i looked at it and realize the fix would only introduce more ugly things i might refrain from doing it however :o)

Whatever, i see some strange behaviour from the makefile now, perhaps you can explain what happens:

- make sure the samples dir is "clean"
- run just "make" or "make SYS=c64"

that works as expected...

now "clean" again, and run for another target, eg "make SYS=apple2" - this will also work, but at the end it shows a message like:

make[1]: „samples“ ist bereits aktuell.

*shrug* what is happening there? :)

**Comment by mrdudz** [comment]
> shrug what is happening there? :)

ok well, got it, fixed :) Enough for this weekend... :)

One thing that might be useful to add to the Makefile in samples would be a target that tests all targets... right now i have a little shell script that calls "make SYS=<target> clean samples" for every target - it would be handy to have that in the makefile itself however (call it something like "testcompile" or so?). perhaps that target can then even be used by the CI instead of whatever it uses now (so really all samples will get compiled, and even with all targets - so they keep working at all times). What do you think?

**Comment by mrdudz** [comment]
PS: now that i read it again... once the makefile is converted to build out of tree, it could very well build into a seperate directory per target - and then we can also have a properly working "all" target

**Comment by mrdudz** [comment]
@oliverschmidt please review the targettest folder and let me know if the current state allows you to take over - will look at the samples folder next

**Comment by mrdudz** [comment]
ok, had some time to go on cleaning this up
- both samples and targettest use the same Makefile logic now and can build everything for any target, including subdirs
- both samples and targettest are now compiled _for every target_ by the github action that triggers on PRs (and it already helped finding a bug: #1436 

additionally to the few programs mentioned above, there is now also a "sym1" subdir in targettest, which probably should also be moved into samples instead.

(and now that everything works fine for automated testing my motivation to work at out of tree building dropped significantly =P)

@oliverschmidt @greg-king5 @groessler please comment on the mentioned programs

**Comment by oliverschmidt** [comment]
> targettest/apple2 - this looks like some sort of diashow program

Actually three independent programs:
* __hgrtest__ shows how to incorporate a hires screen right into the binary, thus allowing to place content both below and above the hires video ram without multiple files nor copying stuff around (plus having a nice splash screen as bonus). It basically demos how to make use of `apple2-hgr.cfg`.
* __hgrshow__ is a (single) hires program showing how to load (single) hires screens from files without support from TGI. TGI could be used to "paint" on the loaded screens.
* __dhgrshow__ is a double hires program showing how to load double hires screens from files. TGI is used for most of the I/O address manipulation, but (in contrast to hgrshow) is not actually functional.

So those three programs are rather samples than test programs.

**Comment by mrdudz** [comment]
remaining things moved, i consider this done now (some Makefile stuff may need refinement - but its all working as intended for now)
